### PR TITLE
Implement MavenResolveTestFixture

### LINKED ISSUE #26826: Add Maven test fixture to test POMs generated by Gradle
After generating a POM, we should verify that Maven consumes it the way we expect.

### PR DESCRIPTION
**Pull request by jvandort**
This test fixture leverages Maven's public APIs for performing dependency resolution with its artifact repository engine. We skip running Maven entirely since that requires downloading a Maven distribution, which itself downloads plugins upon startup. This approach is more flexible for future expansion since it offers the ability to traverse the resolved graph itself, in-process This approach is much faster than running maven in a forked process or running MavenCli embedded.

Fixes #26826

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishExternalVariantIntegrationTest.groovy [hunk 1]
```diff
@@ -861,5 +1044,30 @@ class MavenPublishExternalVariantIntegrationTest extends AbstractMavenPublishInt
...
+                    println "'${name}' classpath: " + allFiles*.name.join(",")
+                }
+            }
+        """
+    }
+
+    Set<String> readClasspath(String name = "") {
+        output.readLines().find { it.startsWith("'${name}' classpath: ") }.split(":")[1].substring(1).split(",") as Set
+    }
+
```
Comment (big-guy) [inline platforms/software/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishExternalVariantIntegrationTest.groovy:1071]: ❓ Can we just write this to a file instead?

**Review by big-guy — CHANGES_REQUESTED** [review]
haven't finished looking at ApiMavenResolver...

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/maven/ApiMavenResolver.groovy [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+     * The local repository used for caching resolved artifacts.
+     * This is the the repository usually located at ~/.m2/repository.
+     */
+    File localRepoDir
+
+    /**
+     * The URI of the repository to resolve artifacts from.
+     * This repository should contain the artifacts under test.
+     */
+    URI testRepoUri
```
[the remark this answers is not in the document]
Comment (jvandort) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/maven/ApiMavenResolver.groovy:59]: These should be final. Probably can make them private too

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/maven/MavenResolveTestFixture.groovy [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * by this fixture to run tests against those published artifacts.</p>
+ */
+@SelfType(AbstractIntegrationSpec.class)
+trait MavenResolveTestFixture {
+
+    /**
+     * Get a resolver that can be used to perform a Maven resolution.
+     */
+    ApiMavenResolver getMavenResolver() {
+        def localRepoDir = temporaryFolder.testDirectory.file("maven-resolver-test-fixture-m2")
```
[the remark this answers is not in the document]
Comment (jvandort) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/maven/MavenResolveTestFixture.groovy:37]: I would argue they should not interact. 
`m2` on the `AbstractIntegSpec` should be used to test how Gradle interacts with `m2`. 
This Maven resolution should be independent from any outside state except for maven central and the repo to test against

File: platforms/software/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishExternalVariantIntegrationTest.groovy (hunk 1 above)
Comment (jvandort) [reply platforms/software/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishExternalVariantIntegrationTest.groovy:1071]: I dont see why not

**Review by big-guy — APPROVED** [review]
I have some ideas for how to improve this, but let's not block 8.5 on it

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Pull request merged**